File: 513-find-bottom-left-tree-value/go/main.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type NodeDepth struct {
	Node  *TreeNode
	Depth int
}

func findBottomLeftValue(root *TreeNode) int {
	return findBottomLeftAndDepth(root, 0).Node.Val
}

func findBottomLeftAndDepth(root *TreeNode, depth int) NodeDepth {
	left := false
	right := false
	leftDepth := 0
	rightDepth := 0
	var leftNode *TreeNode
	var rightNode *TreeNode

	if root.Left != nil {
		left = true
		nd := findBottomLeftAndDepth(root.Left, depth+1)
		leftDepth = nd.Depth
		leftNode = nd.Node
	}

	if root.Right != nil {
		right = true
		nd := findBottomLeftAndDepth(root.Right, depth+1)
		rightDepth = nd.Depth
		rightNode = nd.Node
	}

	if left && right {
		if rightDepth > leftDepth {
			return NodeDepth{Node: rightNode, Depth: rightDepth}
		} else {
			return NodeDepth{Node: leftNode, Depth: leftDepth}
		}
	}

	if left {
		return NodeDepth{Node: leftNode, Depth: leftDepth}
	}

	if right {
		return NodeDepth{Node: rightNode, Depth: rightDepth}
	}

	return NodeDepth{Node: root, Depth: depth}
}

File: 513-find-bottom-left-tree-value/cpp/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
 public:
  int findBottomLeftValue(TreeNode* root) {
    return findLeftMostAndDepth(root, 0).first->val;
  }

  pair<TreeNode*, int> findLeftMostAndDepth(TreeNode* root, int depth) {
    bool left = false;
    bool right = false;
    TreeNode* leftNode = NULL;
    TreeNode* rightNode = NULL;
    int leftDepth = 0;
    int rightDepth = 0;

    if (root->left != NULL) {
      left = true;
      auto p = findLeftMostAndDepth(root->left, depth + 1);
      leftDepth = p.second;
      leftNode = p.first;
    }

    if (root->right != NULL) {
      right = true;
      auto p = findLeftMostAndDepth(root->right, depth + 1);
      rightDepth = p.second;
      rightNode = p.first;
    }

    if (left && right) {
      if (rightDepth > leftDepth) {
        return {rightNode, rightDepth};
      } else {
        return {leftNode, leftDepth};
      }
    }

    if (left) {
      return {leftNode, leftDepth};
    }

    if (right) {
      return {rightNode, rightDepth};
    }

    return {root, depth};
  }
};
